### PR TITLE
Add schema validation to replication worker

### PR DESCRIPTION
**Pull request by alovew**
## What
Update to the DefaultReplicationWorker to use a new RecordSchemaValidation class to validate each AirbyteRecord from the source to ensure that the schema of each record is the same as what is declared in the ConfiguredAirbyteCatalog. For now, we are logging results instead of raising an error when the record schema is invalid.

The RecordSchemaValidator is a new class for validating record message schemas. This class is initialized with a map of source stream names to json schemas, so that when we validate a record, we can easily access the schema it should be validated against.

The record validation happens within the DefaultReplicationWorker before we update the Record Message's attributes with `mapMessage`.

### CODE REVIEW

File: airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java
Comment (lmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java:38]:
```suggestion
      throw new RecordSchemaValidationException(String.format("Record schema validation failed. Errors: %s", e.getMessage()), e);
```
We probably want the json validation exception to be set as the cause, so just moving this parenthesis over to achieve that

File: airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java
Comment (lmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java:82]: We should add a `@VisibleForTesting` annotation to this constructor and remove the `public` keyword to make this [package-private](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html), since we only expect this one to be used for testing

File: airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java
Comment (lmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java:17]: I think it would slightly simplify this interface if this took in an `AirbyteRecordMessage` rather than an `AirbyteMessage`. This is also helpful as it indicates that we only expect to pass _record_ messages to this class

File: airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java
Comment (lmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java:25]: Right now, this logic to read all of the catalog's streams and filter them down in memory happens for every record that is being validated, which could hurt performance when this is done for thousands or millions of records. We also don't expect this `syncInput` to change from one record to the next.

Instead, this `RecordSchemaValidator` class could take in the `StandardSyncInput` in the constructor, and at construction time it could transform the catalog streams into a map of correctly-formatted stream names to the corresponding ConfiguredAirbyteStream object (i.e. apply `streamPrefix + s.getStream().getName().trim()` to each stream to get the key, with the value being the ConfiguredAirbyteStream itself). 

This way, each individual record validation would just need to check if there is a matching key in the map, and retrieve the corresponding value if so, which is a constant-time operation rather than the O(n) filter operation that we have here.

File: airbyte-workers/src/test/java/io/airbyte/workers/DefaultReplicationWorkerTest.java
Comment (lmossman) [inline airbyte-workers/src/test/java/io/airbyte/workers/DefaultReplicationWorkerTest.java:141]: Is there a reason you replaced this test instead of just adding a new one and leaving this one as-is?

File: airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java
Comment (alovew) [reply airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java:25]: I was starting to implement this, but we don't actually have the StandardSyncInput when the RecordSchemaValidator is constructed. It's constructed when the DefaultReplicationWorker is constructed, but the StandardSyncInput is passed into the `run` method. We could move the SyncInput to be passed into the constructor as well - I think in places where we initialize the DefaultReplicationWorker we also already have the SyncInput -- or do you think the recordSchemaValidator should be moved to be passed into the `run` method?

Comment (lmossman) [reply airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java:25]: I think moving the SyncInput to be passed into the DefaultReplicationWorker constructor makes sense - as you said it looks like we need the sync input at construction time anyway in order to initialize things like the NamespacingMapper, so this doesn't seem like a huge lift to me. Plus, the jobId is already something we pass to that constructor, and the syncInput should always be the same for a given job id, so conceptually the DefaultReplicationWorker is already scoped to a single job with a single syncInput.

Comment (lmossman) [reply airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java:25]: In fact, it might make sense for us to just have a single constructor for the DefaultReplicationWorker that takes in the RecordSchemaValidator, and that RecordSchemaValidator takes in the SyncInput in its constructor. This way we are only adding one new constructor parameter to DefaultReplicationWorker, since that is all we would be using the SyncInput for anyway.

Comment (lmossman) [reply airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java:25]: Also one more note - it would probably be slightly more efficient for the map that I described earlier in this comment chain to just store the corresponding stream's `jsonSchema` as the value, rather than storing the entire ConfiguredAirbyteStream as map value, since the jsonSchema is all we really need.

File: airbyte-workers/src/test/java/io/airbyte/workers/DefaultReplicationWorkerTest.java
Comment (alovew) [reply airbyte-workers/src/test/java/io/airbyte/workers/DefaultReplicationWorkerTest.java:141]: oops

File: airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidationException.java
Comment (lmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidationException.java:7]: Could you add a brief javadoc to this class explaining what it means / what can cause this?

File: airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java
Comment (lmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java:27]: I had suggested adding this logic of creating a map of stream name to stream schema in the constructor so that we didn't have to filter the list of streams on every `validateSchema` call, which you did correctly here.

However, looking back on this I realize that we probably should not have somewhat complex logic and the creation of objects like this inside of constructors, since I believe that is sort of an antipattern in Java.

@jdpgrailsdev do you have any thoughts here on how we should approach this? I think there is still value in creating this map of stream name -> stream record schema up front so that the validation calls themselves are very quick, but perhaps this should be done in a factory class/method instead? What is the best approach with the future DI work in mind?

Comment (jdpgrailsdev) [reply airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java:32]: @lmossman I am missing some context here about the `StandardSyncInput`, catalogs and streams (how/where they are created, etc).  Generally speaking, I would agree that in the DI world, this map would be something that we would want to create once and inject as a bean.  Of course, that only works if the map can be created separate from this code/can be constructed at startup.  In the meantime, any change that we could make that moves to a "create once" approach would be best and would make the transition to DI easier.  Maybe this code could be encapsulated in the `StandardSyncInput` so that it is built when that is created and saved as a member and exposed via method.

Sort of related, that lambda could be re-written to use `.collect(Collectors.toMap(k -> String.format(streamPrefix + k.getStream().getName().trim()), v -> v.getStream().getJsonSchema())`

Comment (lmossman) [reply airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java:32]: Following this all the way up the chain, the `StandardSyncInput` is first constructed [here](https://github.com/airbytehq/airbyte/blob/fd7f21e07170fbc9cb6a22d03b642486d0da0b52/airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java#L59-L70), and then is passed down through a bunch of classes until it is eventually serialized and written to a file on the orchestrator pod. This pod then deserializes that file back into an object [here](https://github.com/airbytehq/airbyte/blob/fd7f21e07170fbc9cb6a22d03b642486d0da0b52/airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/ReplicationJobOrchestrator.java#L59), where it is then passed into the DefaultReplicationWorker [here](https://github.com/airbytehq/airbyte/blob/fd7f21e07170fbc9cb6a22d03b642486d0da0b52/airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/ReplicationJobOrchestrator.java#L102). 

Would a good intermediate change be to add this map creation logic to the ReplicationJobOrchestrator after the object is deserialized, and then pass that map into this RecordSchemaValidationException class when it is constructed in that method?

Comment (lmossman) [reply airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java:32]: I think it will be hard to encapsulate this code in `StandardSyncInput`, because that is just a POJO that gets generated from this json schema: https://github.com/airbytehq/airbyte/blob/master/airbyte-config/models/src/main/resources/types/StandardSyncInput.yaml

I think it will be a pretty ugly change to try to add this map all the way up the call chain to the same place that the StandardSyncInput is first constructed, where I linked above. I don't want to block this PR much longer on this decision, so maybe it makes sense to just leave this as-is for now. Though @alovew I think we should rewrite lambda function as Jonathan proposed above, which will allow the `this.streams` initalization to all happen on a single line

Comment (jdpgrailsdev) [reply airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java]: @lmossman The general dependency injection guidance would be that objects should not create object internally.  For instance, if class A currently instantiates an instance of class B in class A's constructor, we would want to change this so that class B is created by whatever common code also instantiates class A and passes the B instance into A (in the future, this would be converted to use auto-wiring).  So, in this case, if there is a higher class that creates the `RecordSchemaValidator` could also create this map and pass it to the `RecordSchemaValidator` via its constructor.  This would allow us to eventually move the creation of the map to the dependency injection configuration, if possible.

Comment (lmossman) [reply airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java]: Got it, then sounds like we should probably go with what I suggested in my first reply to you above and create the map outside of the RecordSchemaValidator, then pass it into `RecordSchemaValidator`. There are currently two different places where the RecordSchemaValidator is constructed (ReplicationJobOrchestrator and ReplicationActivityImpl) so we will probably want to have some sort of utility/library function to create this map from the standard sync input, or a factory class for the ReplicationActivityImpl that does this. I'll leave that decision up to @alovew as the implementer of these chanes

File: airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java
Comment (lmossman) [inline airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java:25]: Since this `syncInput` argument is no longer used by the constructor, it can be removed entirely

**Comment by alovew** [comment]
This is ready for another review. @benmoriceau I added you as a reviewer since Lake is out today and Monday.

File: airbyte-workers/src/main/java/io/airbyte/workers/WorkerUtils.java
Comment (benmoriceau) [inline airbyte-workers/src/main/java/io/airbyte/workers/WorkerUtils.java:106]: Can we extract the strings in var and get rid of the format, I don't see what that brings.

```
k -> {
   String namespace = Objects.toString(k.getStream().getNamespace(), ""));
   String name = k.getStream().getName().trim();

   return namespace + name;
}
```

**Review by benmoriceau — COMMENTED** [review]
Nit comment

**Comment by alovew** [comment]
@benmoriceau @jdpgrailsdev bumping this for review!

**Review by jdpgrailsdev — APPROVED** [review]
:shipit: Code-wise, this looks good to me.  I'll defer to @benmoriceau from a context point of view.

**Review by benmoriceau — APPROVED** [review]

File: airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java
Comment (ChristopheDuong) [inline airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java:298]: @alovew is this printing a warning message for each record during a sync (in case of bad validation)? 

In case of a large sync where all records are badly formatted the same way, this would fill up the logs with tons of similar warning if i am not wrong?

See reports from users here: https://github.com/airbytehq/airbyte/issues/12804

**Pull request merged**
